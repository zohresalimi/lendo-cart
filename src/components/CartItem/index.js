import React, { useContext, useState, useCallback } from "react";
import { Link, useLocation } from "react-router-dom";

import {
  TableRow,
  TableCell,
  TextField,
  Box,
  IconButton,
} from "@material-ui/core";
import ArrowDropUpIcon from "@material-ui/icons/ArrowDropUp";
import ArrowDropDownIcon from "@material-ui/icons/ArrowDropDown";
import { formatPrice } from "../../utils";
import AppContext from "../../store/context";

import {
  INCREASE_QUANTITY,
  DECREASE_QUANTITY,
  REMOVE_ITEM,
} from "../../constant";

import useStyles from "../CartItem/style";
import { RemoveCircleOutline } from "@material-ui/icons";

export default function CartItem({ product }) {
  const { state, dispatch } = useContext(AppContext);
  const { products } = state;

  const remaining =
    products.productByIds[product.id].byColor[product.selectedFeatures.color]
      .remaining;

  const increaseQuantity = useCallback(() => {
    if (remaining !== 0) {
      dispatch({ type: INCREASE_QUANTITY, data: product });
    }
  }, [dispatch, product, remaining]);

  const decreaseQuantity = useCallback(() => {
    if (product.quantity !== 1) {
      dispatch({ type: DECREASE_QUANTITY, data: product });
    }
  }, [dispatch, product]);

  const removeItem = useCallback(() => {
    dispatch({ type: REMOVE_ITEM, data: product });
  }, [dispatch, product]);

  const classes = useStyles();

  return (
    <TableRow key={product.id}>
      <TableCell component="th" scope="row">
        <p>{products.productByIds[product.id].name}</p>

        <p>
          Brand: <span>{products.productByIds[product.id].brand}</span>
        </p>
      </TableCell>
      <TableCell align="right">
        {formatPrice(products.productByIds[product.id].price)}
      </TableCell>
      <TableCell align="right">
        <Box
          display="flex"
          justifyContent="center"
          flexDirection="row"
          alignItems="center"
        >
          <Box>
            <TextField
              className={classes.inputWrapper}
              variant="outlined"
              defaultValue={product.quantity}
              value={product.quantity}
              InputProps={{
                className: classes.input,
              }}
            />
          </Box>
          <Box
            display="flex"
            justifyContent="center"
            flexDirection="column"
            alignItems="center"
            className={classes.btnWrapper}
          >
            <IconButton
              aria-label="plusOne"
              className={classes.plusBtn}
              onClick={increaseQuantity}
              disabled={remaining === 0}
            >
              <ArrowDropUpIcon />
            </IconButton>
            <IconButton
              aria-label="minusOne"
              className={classes.minusBtn}
              onClick={decreaseQuantity}
              disabled={product.quantity <= 1}
            >
              <ArrowDropDownIcon />
            </IconButton>
          </Box>
        </Box>
      </TableCell>
      <TableCell align="right">
        {formatPrice(
          products.productByIds[product.id].price * product.quantity
        )}
      </TableCell>
      <TableCell align="right">
        <IconButton aria-label="remove" onClick={removeItem}>
          <RemoveCircleOutline />
        </IconButton>
      </TableCell>
    </TableRow>
  );
}
