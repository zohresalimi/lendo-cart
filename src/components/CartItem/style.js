import { makeStyles } from "@material-ui/core/styles";
const useStyles = makeStyles((theme) => ({
  inputWrapper: {
    margin: "2px",
    padding: "0",
    maxWidth: "75px",
  },
  input: {
    borderRadius: "0",

    "& input": {
      padding: "16px",
      textAlign: "center",
    },
  },
  btnWrapper: {
    " & button": {
      border: "1px solid #ccc",
      borderRadius: "0",
      padding: "0px",
    },
  },
}));

export default useStyles;
