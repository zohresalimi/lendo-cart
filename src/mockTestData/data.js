import React, { Children, useReducer } from "react";
import { BrowserRouter as Router } from "react-router-dom";

import AppContext from "../store/context";
import reducer from "../store/reducer";

export const getTestStore = () => {
  return {
    currentProduct: {
      id: 1,
      selectedFeatures: {
        color: "whith",
        power: 6.5,
      },
    },
    products: {
      byColor: {
        white: {
          1: { power: [6.5, 9.5], remaining: 3 },
          2: { power: [6.5, 9.5], remaining: 3 },
        },
        red: {
          1: { power: [6.5, 9.5], remaining: 7 },
          2: { power: [6.5, 9.5], remaining: 7 },
        },
      },
      byId: {
        1: {
          id: 1,
          name: "Philips hue bulb",
          brand: "Philips",
          price: "500",
          available: true,
          weight: 0.2,
          options: [
            {
              color: "white",
              power: [6.5, 9.5],
              quantity: 3,
            },
            {
              color: "red",
              power: [6.5, 9.5],
              quantity: 7,
            },
          ],
        },
        2: {
          id: 2,
          name: "TrÃ¥dfria Lampor",
          brand: "IKEA",
          price: "300",
          available: true,
          weight: 0.2,
          options: [
            {
              color: "white",
              power: [6.5, 9.5],
              quantity: 3,
            },
            {
              color: "red",
              power: [6.5, 9.5],
              quantity: 7,
            },
            {
              color: "green",
              power: [6.5],
              quantity: 0,
            },
          ],
        },
      },
    },
    shoppingCart: [
      { id: 1, quantity: 1, selectedFeatures: { color: "white", power: 6.5 } },
    ],
  };
};

export const WithProvider = (props) => {
  const [state, dispatch] = useReducer(reducer, getTestStore());

  return (
    <AppContext.Provider
      value={{
        state,
        dispatch: props.mockDispatch || dispatch,
      }}
    >
      <Router>{props.children}</Router>
    </AppContext.Provider>
  );
};
