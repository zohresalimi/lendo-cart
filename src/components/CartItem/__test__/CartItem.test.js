import * as React from "react";

import { act, render, waitFor, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import { WithProvider, getTestStore } from "../../../mockTestData/data";

import CartItem from "..";

async function renderWrapper(props, mockDispatch) {
  let component;

  act(() => {
    component = render(
      <WithProvider mockDispatch={mockDispatch}>
        <CartItem {...props} />
      </WithProvider>,
      {
        container: document.createElement("tbody"),
      }
    );
  });

  return component;
}

describe("Cart Item", () => {
  test("should match snapshot", async () => {
    const testState = getTestStore();
    const { container } = await renderWrapper({
      product: testState.shoppingCart[0],
      onRemove: () => {},
    });

    expect(container).toMatchSnapshot();
  });

  test("should not decrease on down-button click if quantity is 1", async () => {
    const testState = getTestStore();
    const mockDispatch = jest.fn();
    const { getByTestId } = await renderWrapper(
      {
        product: testState.shoppingCart[0],
        onRemove: () => {},
      },
      mockDispatch
    );

    await waitFor(() => {
      const decreaseButton = getByTestId("decrease-button");

      fireEvent.click(decreaseButton);
      expect(mockDispatch).toHaveBeenCalledTimes(0);
    });
  });

  test("should decrease quantity on down-button click", async () => {
    const testState = getTestStore();
    const mockDispatch = jest.fn();
    const { getByTestId } = await renderWrapper(
      {
        product: {
          ...testState.shoppingCart[0],
          quantity: 2,
        },
        onRemove: () => {},
      },
      mockDispatch
    );

    await waitFor(() => {
      const decreaseButton = getByTestId("decrease-button");

      fireEvent.click(decreaseButton);
      expect(mockDispatch).toHaveBeenLastCalledWith({
        type: "decreaseQuantity",
        data: {
          ...testState.shoppingCart[0],
          quantity: 2,
        },
      });
    });
  });

  test("should increase quantity on up-button click", async () => {
    const testState = getTestStore();
    const mockDispatch = jest.fn();
    const { getByTestId } = await renderWrapper(
      {
        product: testState.shoppingCart[0],
        onRemove: () => {},
      },
      mockDispatch
    );

    await waitFor(() => {
      const increaseButton = getByTestId("increase-button");

      fireEvent.click(increaseButton);
      expect(mockDispatch).toHaveBeenCalledWith({
        type: "increaseQuantity",
        data: testState.shoppingCart[0],
      });
    });
  });

  test("should remove item on remove-button click", async () => {
    const testState = getTestStore();
    const mockDispatch = jest.fn();
    const mockRemoveCallback = jest.fn();
    const { getByTestId } = await renderWrapper(
      {
        product: testState.shoppingCart[0],
        onRemove: mockRemoveCallback,
      },
      mockDispatch
    );

    await waitFor(() => {
      const removeButton = getByTestId("remove-button");

      fireEvent.click(removeButton);
      expect(mockRemoveCallback).toHaveBeenCalled();
      expect(mockDispatch).toHaveBeenCalledWith({
        type: "removeItem",
        data: testState.shoppingCart[0],
      });
    });
  });
});
