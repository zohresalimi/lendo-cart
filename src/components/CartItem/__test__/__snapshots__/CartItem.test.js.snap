// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart Item should match snapshot 1`] = `
<tbody>
  <tr
    class="MuiTableRow-root"
  >
    <th
      class="MuiTableCell-root"
      role="cell"
      scope="row"
    >
      <p>
        Philips hue bulb
      </p>
      <p>
        Brand: 
        <span>
          Philips
        </span>
      </p>
      <div
        class="MuiBox-root MuiBox-root-4"
      >
        color
        : 
        white
      </div>
      <div
        class="MuiBox-root MuiBox-root-5"
      >
        power
        : 
        6.5
      </div>
    </th>
    <td
      class="MuiTableCell-root MuiTableCell-alignRight"
    >
      500,00 kr
    </td>
    <td
      class="MuiTableCell-root MuiTableCell-alignRight"
    >
      <div
        class="MuiBox-root MuiBox-root-6"
      >
        <div
          class="MuiBox-root MuiBox-root-7"
        >
          <div
            class="MuiFormControl-root MuiTextField-root makeStyles-inputWrapper-1"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root makeStyles-input-2 MuiInputBase-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input"
                type="text"
                value="1"
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-8 MuiOutlinedInput-notchedOutline"
                style="padding-left: 8px;"
              >
                <legend
                  class="PrivateNotchedOutline-legend-9"
                  style="width: 0.01px;"
                >
                  <span>
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-12 makeStyles-btnWrapper-3"
        >
          <button
            aria-label="plusOne"
            class="MuiButtonBase-root MuiIconButton-root"
            data-testid="increase-button"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 14l5-5 5 5z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            aria-label="minusOne"
            class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
            data-testid="decrease-button"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </td>
    <td
      class="MuiTableCell-root MuiTableCell-alignRight"
    >
      500,00 kr
    </td>
    <td
      class="MuiTableCell-root MuiTableCell-alignRight"
    >
      <button
        aria-label="remove"
        class="MuiButtonBase-root MuiIconButton-root"
        data-testid="remove-button"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 11v2h10v-2H7zm5-9C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </td>
  </tr>
</tbody>
`;
